app/src/main/java/com/example/newsapp/features/auth/register/RegisterViewModel.kt

```diff
@@ -0,0 +1,86 @@
+@file: Suppress("WildcardImport", "NoWildcardImports")
+
+package com.example.newsapp.features.auth.register
+
+import androidx.lifecycle.*
+import com.example.newsapp.data.entities.UserDto
+import com.example.newsapp.data.repos.UserRepository
+import com.example.newsapp.utils.isValidEmailFormat
+import com.example.newsapp.utils.isValidPasswordFormat
+import com.example.newsapp.utils.isValidPhoneNumberFormat
+import com.example.newsapp.utils.toEntity
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+class RegisterViewModel(
+    private val userRepository: UserRepository,
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
+) : ViewModel() {
+    private val registrationResultMutableLiveData = MutableLiveData<RegisterResult>()
+    val registrationResultLiveData: LiveData<RegisterResult>
+        get() = registrationResultMutableLiveData
+
+    fun register(user: UserDto) {
+        if (user.name.isEmpty()) {
+            registrationResultMutableLiveData.value =
+                RegisterResult.InvalidData(ErrorType.EmptyName)
+        } else if (user.email.isEmpty()) {
+            registrationResultMutableLiveData.value =
+                RegisterResult.InvalidData(ErrorType.EmptyEmail)
+        } else if (!user.email.isValidEmailFormat()) {
+            registrationResultMutableLiveData.value =
+                RegisterResult.InvalidData(ErrorType.InvalidEmailFormat)
+        } else if (user.password.isEmpty()) {
+            registrationResultMutableLiveData.value =
+                RegisterResult.InvalidData(ErrorType.EmptyPassword)
+        } else if (!user.password.isValidPasswordFormat()) {
+            registrationResultMutableLiveData.value =
+                RegisterResult.InvalidData(ErrorType.InvalidPasswordFormat)
+        } else if (user.phoneNumber.isEmpty()) {
+            registrationResultMutableLiveData.value =
+                RegisterResult.InvalidData(ErrorType.EmptyPhoneNumber)
+        } else if (!user.phoneNumber.isValidPhoneNumberFormat()) {
+            registrationResultMutableLiveData.value =
+                RegisterResult.InvalidData(ErrorType.InvalidPhoneNumberFormat)
+        } else {
+            // all data is valid
+            viewModelScope.launch(ioDispatcher) {
+                when (userRepository.registerUser(user.toEntity())) {
+                    true -> {
+                        registrationResultMutableLiveData.postValue(RegisterResult.RegisterSuccessful)
+                    }
+                    false -> {
+                        registrationResultMutableLiveData.postValue(RegisterResult.RegisterError)
+                    }
+                }
+            }
+        }
+    }
+}
+
+sealed class RegisterResult {
+    object RegisterSuccessful : RegisterResult()
+    object RegisterError : RegisterResult()
+    data class InvalidData(val error: ErrorType) : RegisterResult()
+}
+
+sealed class ErrorType {
```

Comment:
I think Enum is better. create an enum that take a string in the constructor for Example `EmptyEmail("Email can't be empty")` this will reduce the branching in the fragment :
```
when (result) {
                is RegisterResult.InvalidData -> {
                      binding.emailET.error = result.value.message // not sure the correct syntax here, I am so lazy to try it.
                 }
```